AMM/contracts/exchange/src/main.sw

```diff
@@ -0,0 +1,423 @@
+        let asset_a_in_reserve = storage.reserves.get(asset_a_id);
+        let asset_b_in_reserve = storage.reserves.get(asset_b_id);
+        let mut added_a = 0;
+        let mut added_b = 0;
+        let mut liquidity_to_add = 0;
+        require(asset_a_in_deposit != 0, TransactionError::DepositCannotBeZero);
+        require(asset_b_in_deposit != 0, TransactionError::DepositCannotBeZero);
+        if asset_a_in_reserve == 0 && asset_b_in_reserve == 0 {
+            liquidity_to_add = (asset_a_in_deposit * asset_b_in_deposit).sqrt();
+            require(desired_liquidity <= liquidity_to_add, TransactionError::DesiredAmountTooHigh(desired_liquidity));
+            added_a = asset_a_in_deposit;
+            added_b = asset_b_in_deposit;
+            storage.reserves.insert(asset_a_id, added_a);
+            storage.reserves.insert(asset_b_id, added_b);
+            mint(liquidity_to_add);
+            storage.liquidity_pool_supply = liquidity_to_add;
+            transfer(liquidity_to_add, contract_id(), sender);
+        } else {
+            let b_to_attempt = multiply_div(asset_a_in_deposit, asset_b_in_reserve, asset_a_in_reserve);
+            if b_to_attempt <= asset_b_in_deposit {
+                liquidity_to_add = multiply_div(b_to_attempt, total_liquidity, asset_b_in_reserve);
+                require(desired_liquidity <= liquidity_to_add, TransactionError::DesiredAmountTooHigh(desired_liquidity));
+                added_a = asset_a_in_deposit;
+                added_b = b_to_attempt;
+            } else {
+                let a_to_attempt = multiply_div(asset_b_in_deposit, asset_a_in_reserve, asset_b_in_reserve);
+                liquidity_to_add = multiply_div(a_to_attempt, total_liquidity, asset_a_in_reserve);
+                require(desired_liquidity <= liquidity_to_add, TransactionError::DesiredAmountTooHigh(desired_liquidity));
+                added_a = a_to_attempt;
+                added_b = asset_b_in_deposit;
+            }
+            storage.reserves.insert(asset_a_id, asset_a_in_reserve + added_a);
+            storage.reserves.insert(asset_b_id, asset_b_in_reserve + added_b);
+            mint(liquidity_to_add);
+            storage.liquidity_pool_supply = total_liquidity + liquidity_to_add;
+            transfer(liquidity_to_add, contract_id(), sender);
+
+            let refund_a = asset_a_in_deposit - added_a;
+            if refund_a > 0 {
+                transfer(refund_a, asset_a_id, sender);
+            }
+            let refund_b = asset_b_in_deposit - added_b;
+            if refund_b > 0 {
+                transfer(refund_b, asset_b_id, sender);
+            }
+        }
+        storage.deposits.insert((sender, asset_a_id), 0);
+        storage.deposits.insert((sender, asset_b_id), 0);
+        log(AddLiquidityEvent {
+            a: added_a,
+            b: added_b,
+            liquidity: liquidity_to_add,
+        });
+        liquidity_to_add
```

Comment:
Hmm. The `preview_add_liquidity` function gives the user a way to compute the liquidity amount to add by providing a specific amount of an asset. But it is by no means a prerequisite to call the `add_liquidity` function. Since at the time of calling the `add_liquidity` function, the amount of liquidity being added is unknown, I think it is useful to return the added liquidity amount.